File: packages/engine/src/core/shipwatch.ts
import { readIndexFile, writeIndexFile } from './index-store'

/**
 * Per-session ref-movement marks — the gate that makes the live shipping
 * signal affordable (commit-attribution 3.4, D11).
 *
 * The problem 3.2 left open: the SessionStart notice fires ONCE, so a window
 * already running when a sibling pushes learns nothing until it restarts. That
 * is the user's original complaint, still open in exactly the case they
 * described. Answering it per prompt means asking "did MY commits reach
 * origin", and that question needs a trailer walk — a subprocess D6 forbids on
 * the hot path outright.
 *
 * The escape D6 was written with is that the expensive answer only CHANGES
 * when a ref moves, and detecting movement is free: origin/<branch> is a file,
 * already read by readGitState with no subprocess. So the hot path compares one
 * sha against the last one this session saw, and the walk is paid only on an
 * actual push. Steady state is a small JSON read and nothing else.
 *
 * The mark lives in the derived index, which is the right home by its own three
 * rules (record-index D1): disposable — losing it costs one missed line, never
 * a wrong one; local, never committed — a mark is about one session on one
 * machine and would conflict on every parallel append; and self-ignoring, so no
 * repo .gitignore has to know. It carries its own version rather than riding
 * INDEX_SCHEMA_VERSION, so a shape change here cold-starts THIS file instead of
 * forcing a full index rebuild.
 *
 * EDGE-TRIGGERED, unlike every other line on that path. The drift nudge and the
 * conflict lines re-fire statelessly because they restate a condition that is
 * still true; this one reports a TRANSITION, and "just landed" repeated ten
 * prompts later is simply false. Marking is therefore a write, and the write is
 * what stops the repeat.
 *
 * CONCURRENCY: sessions share one worktree here, so they share this file. The
 * write is atomic, but a read-modify-write race between two prompts in the same
 * instant REVERTS the loser's mark rather than dropping it — the winner writes
 * back the copy it read, which still carries the loser's OLD sha. So the loser
 * does not read as a first look (silent); it re-detects the same movement and
 * announces the push a second time. Stated plainly because the first version of
 * this comment claimed the opposite, and the difference is the whole
 * edge-trigger property: the race costs a DUPLICATE line, never a missed one,
 * and never a wrong attribution, since a mark is only ever replaced by a sha
 * read from the same refs.
 */

/** Filename inside the index dir. Version lives in the payload, not the name. */
export const SHIPWATCH_FILE = 'shipwatch.json'

/** Bump on ANY change to the on-disk shape — a mismatch cold-starts. */
export const SHIPWATCH_VERSION = 1

/**
 * Marks kept before the oldest are evicted. Sessions are never cleaned up on
 * end (a crashed session ends nothing), so without a cap this grows for the
 * life of the repo. Generous enough that every plausibly-live window keeps its
 * mark; small enough that the file stays a few KB on the hot path.
 */
export const SHIPWATCH_MAX_MARKS = 64

export interface WatchMark {
  /** Branch the mark was taken on — a switch invalidates it (different ref). */
  branch: string
  /**
   * Full sha of origin/<branch> when this session last looked, or null when the
   * ref did not exist then.
   *
   * Null is a WATCHED state, not an unwatchable one, and getting that wrong is
   * what made the first version miss its own headline case: a branch's first
   * push creates the ref, so treating "no upstream" as nothing to watch made
   * the most unambiguous shipping event of all — the work leaving the machine
   * for the first time — the one event that could never be reported.
   */
  upstream: string | null
  /**
   * Write order — the highest on disk plus one. Eviction only; never compared
   * against event time or anything outside this file.
   *
   * A counter rather than a clock, because eviction has to be a total order and
   * a clock is not one here: several marks routinely land inside the same
   * millisecond, and sorting on a tie makes eviction arbitrary — a fresh mark
   * can lose its place to one that has been stale for hours. Deriving the next
   * value from the file also keeps the module free of any clock at all.
   */
  seq: number
}

interface WatchDisk {
  version: number
  marks: Record<string, WatchMark>
}

export interface WatchLook {
  /**
   * The sha this session last saw on this branch. Null carries two different
   * meanings, and `moved` is what separates them: with `moved` false there was
   * no usable mark at all (a first look, a branch switch, an evicted entry),
   * and with `moved` true the ref genuinely did not exist last time — so the
   * range to walk is everything reachable from the new ref, not a delta.
   */
  previous: string | null
  /** True only when a usable mark existed AND the ref differs from it. */
  moved: boolean
}

function isWatchDisk(v: unknown): v is WatchDisk {
  if (typeof v !== 'object' || v === null) return false
  const disk = v as Partial<WatchDisk>
  if (disk.version !== SHIPWATCH_VERSION) return false
  return typeof disk.marks === 'object' && disk.marks !== null
}

const FULL_SHA = /^[0-9a-f]{40}$/

function readMarks(sofarDir: string): Record<string, WatchMark> {
  const disk = readIndexFile<WatchDisk>(sofarDir, SHIPWATCH_FILE, isWatchDisk)
  if (disk === null) return {}
  const clean: Record<string, WatchMark> = {}
  for (const [id, mark] of Object.entries(disk.marks)) {
    if (typeof mark !== 'object' || mark === null) continue
    const m = mark as Partial<WatchMark>
    if (typeof m.branch !== 'string' || m.branch.length === 0) continue
    if (m.upstream !== null && (typeof m.upstream !== 'string' || !FULL_SHA.test(m.upstream))) continue
    if (typeof m.seq !== 'number' || !Number.isFinite(m.seq)) continue
    clean[id] = { branch: m.branch, upstream: m.upstream, seq: m.seq }
  }
  return clean
}

/**
 * Record where origin/<branch> stands for this session, and report what moved
 * since it last looked.
 *
 * Compare-and-mark in one call rather than a read plus a write, because the two
 * halves must see the same file: reading the mark and then writing it back from
 * a second read is how a session announces the same push twice.
 *
 * Writes on every look, not only on change — see the note at the write itself:
 * ordering marks by last LOOK is what stops eviction starving a quiet session.
 *
 * Best-effort throughout — the index writer swallows its own failures — and the
 * two shapes that failure takes are worth stating exactly, because they are
 * opposites. With no mark ever persisted (an unwritable index dir), every look
 * reads as a first look and the caller stays permanently SILENT. With a mark on
 * disk that can no longer be updated, the same movement is re-detected every
 * prompt and the caller REPEATS. Silence is the safe direction and the likelier
 * one; the repeat is the annoying one, and it is bounded by the fact that
 * anything able to write the index again resynchronises on the next look.
 * Neither shape can produce a WRONG attribution, which is the property that
 * matters: a mark is only ever replaced by a sha read from the same refs.
 */
export function noteUpstream(
  sofarDir: string,
  sessionId: string,
  branch: string,
  upstream: string | null,
): WatchLook {
  // A short sha is a display value; accepting one would store a prefix that
  // can go ambiguous later, and an ambiguous rev makes the range error out.
  if (sessionId.length === 0 || branch.length === 0) return { previous: null, moved: false }
  if (upstream !== null && !FULL_SHA.test(upstream)) return { previous: null, moved: false }

  const marks = readMarks(sofarDir)
  const prior = marks[sessionId]
  // A mark taken on another branch describes a different ref entirely, so it
  // is no evidence about this one — treated as a first look, never as movement.
  const usable = prior !== undefined && prior.branch === branch ? prior : undefined

  // Refresh on EVERY look, even when nothing moved, so `seq` orders marks by
  // last LOOK rather than last WRITE. Write-order alone starves the quiet
  // session: one that has seen no push never rewrites, its seq freezes, and
  // SHIPWATCH_MAX_MARKS writes by busier sessions evict a window that is still
  // live — so the push it was waiting for is precisely the one it never hears,
  // its next look reading as a first look and staying silent. That defeats the
  // cap's stated guarantee, and the extra cost is one small write per prompt.
  marks[sessionId] = { branch, upstream, seq: nextSeq(marks) }
  writeIndexFile(sofarDir, SHIPWATCH_FILE, { version: SHIPWATCH_VERSION, marks: evict(marks) })

  if (usable === undefined) return { previous: null, moved: false }
  return usable.upstream === upstream
    ? { previous: upstream, moved: false }
    : { previous: usable.upstream, moved: true }
}

/** One past the highest write order on disk — never reused, never a clock. */
function nextSeq(marks: Record<string, WatchMark>): number {
  let max = 0
  for (const mark of Object.values(marks)) if (mark.seq > max) max = mark.seq
  return max + 1
}

/** Keep the SHIPWATCH_MAX_MARKS most recently written marks; drop the rest. */
function evict(marks: Record<string, WatchMark>): Record<string, WatchMark> {
  const ids = Object.keys(marks)
  if (ids.length <= SHIPWATCH_MAX_MARKS) return marks
  const kept: Record<string, WatchMark> = {}
  for (const id of ids.sort((a, b) => marks[b]!.seq - marks[a]!.seq).slice(0, SHIPWATCH_MAX_MARKS)) {
    kept[id] = marks[id]!
  }
  return kept
}

File: crates/sofar-core/src/shipwatch.rs
//! Per-session ref-movement marks (`core/shipwatch.ts`, commit-attribution
//! 3.4, D11): `shipwatch.json` in the index dir, version 2, ≤64 marks by `seq`.

use crate::index_store::{read_index_file, write_index_file};
use crate::json::{Json, Object};
use crate::layout::Layout;

pub const SHIPWATCH_FILE: &str = "shipwatch.json";
pub const SHIPWATCH_VERSION: f64 = 2.0;
pub const SHIPWATCH_MAX_MARKS: usize = 64;

#[derive(Debug, Clone, PartialEq)]
pub struct WatchMark {
    pub branch: String,
    pub upstream: Option<String>,
    pub engine: Option<String>,
    pub seq: f64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct WatchLook<'a> {
    pub previous: Option<&'a str>,
    pub moved: bool,
}

fn is_full_sha(s: &str) -> bool {
    s.len() == 40 && s.bytes().all(|b| matches!(b, b'0'..=b'9' | b'a'..=b'f'))
}

/// `readMarks`: the clean marks, in file order.
fn read_marks(layout: &Layout) -> Vec<(String, WatchMark)> {
    let Some(disk) = read_index_file(layout, SHIPWATCH_FILE) else {
        return Vec::new();
    };
    let Some(o) = disk.as_obj() else {
        return Vec::new();
    };
    if o.get("version") != Some(&Json::Num(SHIPWATCH_VERSION)) {
        return Vec::new();
    }
    let Some(marks) = o.get("marks").and_then(Json::as_obj) else {
        return Vec::new();
    };
    let mut clean = Vec::new();
    for (id, mark) in marks.js_ordered() {
        let Some(m) = mark.as_obj() else { continue };
        let Some(branch) = m.get("branch").and_then(Json::as_str) else {
            continue;
        };
        let upstream = match m.get("upstream") {
            Some(Json::Null) => None,
            Some(Json::Str(s)) if is_full_sha(s) => Some(s.clone()),
            _ => continue,
        };
        let Some(seq) = m
            .get("seq")
            .and_then(Json::as_f64)
            .filter(|s| s.is_finite())
        else {
            continue;
        };
        let engine = m
            .get("engine")
            .and_then(Json::as_nonempty_str)
            .map(str::to_owned);
        clean.push((
            id.to_owned(),
            WatchMark {
                branch: branch.to_owned(),
                upstream,
                engine,
                seq,
            },
        ));
    }
    clean
}

fn next_seq(marks: &[(String, WatchMark)]) -> f64 {
    marks.iter().map(|(_, m)| m.seq).fold(0.0, f64::max) + 1.0
}

fn evict(mut marks: Vec<(String, WatchMark)>) -> Vec<(String, WatchMark)> {
    if marks.len() <= SHIPWATCH_MAX_MARKS {
        return marks;
    }
    // `ids.sort((a, b) => marks[b].seq - marks[a].seq)` — stable, seq descending.
    marks.sort_by(|a, b| {
        b.1.seq
            .partial_cmp(&a.1.seq)
            .unwrap_or(std::cmp::Ordering::Equal)
    });
    marks.truncate(SHIPWATCH_MAX_MARKS);
    marks
}

fn write_marks(layout: &Layout, marks: Vec<(String, WatchMark)>) {
    let mut obj = Object::with_capacity(marks.len());
    for (id, m) in evict(marks) {
        let mut mo = Object::with_capacity(4);
        mo.insert("branch", Json::Str(m.branch));
        mo.insert("upstream", m.upstream.map_or(Json::Null, Json::Str));
        mo.insert("seq", Json::Num(m.seq));
        if let Some(engine) = m.engine {
            mo.insert("engine", Json::Str(engine));
        }
        obj.insert(id, Json::Obj(mo));
    }
    let mut o = Object::with_capacity(2);
    o.insert("version", Json::Num(SHIPWATCH_VERSION));
    o.insert("marks", Json::Obj(obj));
    write_index_file(layout, SHIPWATCH_FILE, &Json::Obj(o));
}

fn set_mark(marks: &mut Vec<(String, WatchMark)>, id: &str, mark: WatchMark) {
    match marks.iter_mut().find(|(i, _)| i == id) {
        Some(slot) => slot.1 = mark,
        None => marks.push((id.to_owned(), mark)),
    }
}

/// `noteUpstream`: record where `origin/<branch>` stands for this session and
/// report what moved since it last looked. Writes on every look.
pub fn note_upstream(
    layout: &Layout,
    session_id: &str,
    branch: &str,
    upstream: Option<&str>,
) -> (Option<String>, bool) {
    if session_id.is_empty() || branch.is_empty() {
        return (None, false);
    }
    if upstream.is_some_and(|u| !is_full_sha(u)) {
        return (None, false);
    }
    let mut marks = read_marks(layout);
    let prior = marks
        .iter()
        .find(|(i, _)| i == session_id)
        .map(|(_, m)| m.clone());
    let usable = prior.as_ref().filter(|p| p.branch == branch).cloned();
    let seq = next_seq(&marks);
    set_mark(
        &mut marks,
        session_id,
        WatchMark {
            branch: branch.to_owned(),
            upstream: upstream.map(str::to_owned),
            engine: prior.and_then(|p| p.engine),
            seq,
        },
    );
    write_marks(layout, marks);
    match usable {
        None => (None, false),
        Some(u) if u.upstream.as_deref() == upstream => (upstream.map(str::to_owned), false),
        Some(u) => (u.upstream, true),
    }
}

/// `noteEngine`: mark the engine this session runs against; the version it saw
/// last when that differs.
pub fn note_engine(layout: &Layout, session_id: &str, engine: &str) -> Option<String> {
    if session_id.is_empty() || engine.is_empty() {
        return None;
    }
    let mut marks = read_marks(layout);
    let prior = marks
        .iter()
        .find(|(i, _)| i == session_id)
        .map(|(_, m)| m.clone());
    let was = prior
        .as_ref()
        .and_then(|p| p.engine.as_deref())
        .filter(|e| *e != engine)
        .map(str::to_owned);
    if prior.as_ref().and_then(|p| p.engine.as_deref()) == Some(engine) {
        return None;
    }
    let seq = next_seq(&marks);
    set_mark(
        &mut marks,
        session_id,
        WatchMark {
            branch: prior.as_ref().map(|p| p.branch.clone()).unwrap_or_default(),
            upstream: prior.and_then(|p| p.upstream),
            engine: Some(engine.to_owned()),
            seq,
        },
    );
    write_marks(layout, marks);
    was
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn marks_are_edge_triggered_and_written_on_every_look() {
        let dir = crate::testing::scratch_dir("shipwatch");
        let layout = Layout::new(&dir);
        let sha_a = "a".repeat(40);
        let sha_b = "b".repeat(40);
        assert_eq!(note_upstream(&layout, "s", "main", None), (None, false));
        assert_eq!(
            std::fs::read_to_string(layout.index_dir().join(SHIPWATCH_FILE)).unwrap(),
            "{\"version\":2,\"marks\":{\"s\":{\"branch\":\"main\",\"upstream\":null,\"seq\":1}}}\n"
        );
        assert_eq!(
            note_upstream(&layout, "s", "main", Some(&sha_a)),
            (None, true)
        );
        assert_eq!(
            note_upstream(&layout, "s", "main", Some(&sha_a)),
            (Some(sha_a.clone()), false)
        );
        assert_eq!(
            note_upstream(&layout, "s", "main", Some(&sha_b)),
            (Some(sha_a.clone()), true)
        );
        assert_eq!(
            note_upstream(&layout, "s", "other", Some(&sha_b)),
            (None, false)
        );
        assert_eq!(note_engine(&layout, "s", "1.0.0"), None);
        assert_eq!(note_engine(&layout, "s", "1.0.0"), None);
        assert_eq!(note_engine(&layout, "s", "1.1.0"), Some("1.0.0".into()));
        std::fs::remove_dir_all(&dir).unwrap();
    }
}
